// @cloudbase/auth/dist/cjs/providers/weixinAuthProvider.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeixinAuthProvider = void 0;
var base_1 = require("./base");
var utilities_1 = require("@cloudbase/utilities/");
var __1 = require("..");
var constants_1 = require("../constants");
var getSdkName = utilities_1.constants.getSdkName, ERRORS = utilities_1.constants.ERRORS, COMMUNITY_SITE_URL = utilities_1.constants.COMMUNITY_SITE_URL;
var RUNTIME = utilities_1.adapters.RUNTIME;
var getQuery = utilities_1.utils.getQuery, getHash = utilities_1.utils.getHash, removeParam = utilities_1.utils.removeParam, printWarn = utilities_1.utils.printWarn;
var catchErrorsDecorator = utilities_1.helpers.catchErrorsDecorator;
var WeixinAuthProvider = exports.WeixinAuthProvider = (function (_super) {
    __extends(WeixinAuthProvider, _super);
    function WeixinAuthProvider(config, appid, scope, state) {
        var _this = _super.call(this, config) || this;
        _this._runtime = config.runtime;
        _this._appid = appid;
        _this._scope = scope;
        _this._state = state || 'weixin';
        return _this;
    }
    WeixinAuthProvider.prototype.signIn = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, printWarn(ERRORS.OPERATION_FAIL, 'API signIn has been deprecated, please use signInWithRedirect insteed')];
            });
        });
    };
    WeixinAuthProvider.prototype.signInWithRedirect = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this._redirect()];
            });
        });
    };
    WeixinAuthProvider.prototype.getRedirectResult = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var code;
            return __generator(this, function (_a) {
                code = getWeixinCode();
                if (!code) {
                    return [2, null];
                }
                return [2, this._signInWithCode(code, options)];
            });
        });
    };
    WeixinAuthProvider.prototype.getLinkRedirectResult = function (options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var withUnionId, code, appid, loginType, hybridMiniapp;
            var _a;
            return __generator(this, function (_b) {
                withUnionId = (_a = options.withUnionId, _a === void 0 ? false : _a);
                code = getWeixinCode();
                if (!code) {
                    return [2, null];
                }
                appid = this._appid;
                loginType = (function (scope) {
                    switch (scope) {
                        case 'snsapi_login':
                            return 'WECHAT-OPEN';
                        default:
                            return 'WECHAT-PUBLIC';
                    }
                })(this._scope);
                hybridMiniapp = this._runtime === RUNTIME.WX_MP ? '1' : '0';
                return [2, this._request.send('auth.linkWithWeixinCode', {
                        payload: {
                            appid: appid,
                            loginType: loginType,
                            code: code,
                            hybridMiniapp: hybridMiniapp,
                            withUnionId: withUnionId,
                        },
                    })];
            });
        });
    };
    WeixinAuthProvider.prototype._redirect = function () {
        var currUrl = removeParam('code', location.href);
        currUrl = removeParam('state', currUrl);
        currUrl = encodeURIComponent(currUrl);
        var host = '//open.weixin.qq.com/connect/oauth2/authorize';
        if (this._scope === 'snsapi_login') {
            host = '//open.weixin.qq.com/connect/qrconnect';
        }
        try {
            location.href = "".concat(host, "?appid=").concat(this._appid, "&redirect_uri=").concat(currUrl, "&response_type=code&scope=").concat(this._scope, "&state=").concat(this._state, "#wechat_redirect");
        }
        catch (e) {
            throw new Error("[".concat(getSdkName(), "][").concat(ERRORS.UNKOWN_ERROR, "]").concat(e));
        }
    };
    WeixinAuthProvider.prototype._signInWithCode = function (code, options) {
        return __awaiter(this, void 0, void 0, function () {
            var accessTokenKey, accessTokenExpireKey, refreshTokenKey, loginType, refreshTokenRes, refreshToken, loginState;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        accessTokenKey = (_a = this._cache.keys, _a.accessTokenKey), accessTokenExpireKey = _a.accessTokenExpireKey, refreshTokenKey = _a.refreshTokenKey;
                        loginType = (function (scope) {
                            switch (scope) {
                                case 'snsapi_login':
                                    return 'WECHAT-OPEN';
                                default:
                                    return 'WECHAT-PUBLIC';
                            }
                        })(this._scope);
                        return [4, this._getRefreshTokenByWXCode(this._appid, loginType, code, options)];
                    case 1:
                        refreshTokenRes = _b.sent();
                        refreshToken = refreshTokenRes.refreshToken;
                        return [4, this._cache.setStoreAsync(refreshTokenKey, refreshToken)];
                    case 2:
                        _b.sent();
                        if (!refreshTokenRes.accessToken) return [3, 4];
                        return [4, this._cache.setStoreAsync(accessTokenKey, refreshTokenRes.accessToken)];
                    case 3:
                        _b.sent();
                        _b.label = 4;
                    case 4:
                        if (!refreshTokenRes.accessTokenExpire) return [3, 6];
                        return [4, this._cache.setStoreAsync(accessTokenExpireKey, String(refreshTokenRes.accessTokenExpire + Date.now()))];
                    case 5:
                        _b.sent();
                        _b.label = 6;
                    case 6:
                        __1.eventBus.fire(__1.EVENTS.LOGIN_STATE_CHANGED);
                        __1.eventBus.fire(__1.EVENTS.LOGIN_TYPE_CHANGED, {
                            env: this._config.env,
                            loginType: constants_1.LOGINTYPE.WECHAT,
                            persistence: this._config.persistence,
                        });
                        return [4, this.refreshUserInfo()];
                    case 7:
                        _b.sent();
                        loginState = new __1.LoginState({
                            envId: this._config.env,
                            cache: this._cache,
                            request: this._request,
                        });
                        return [4, loginState.checkLocalStateAsync()];
                    case 8:
                        _b.sent();
                        return [2, loginState];
                }
            });
        });
    };
    WeixinAuthProvider.prototype._getRefreshTokenByWXCode = function (appid, loginType, code, options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var withUnionId, createUser, syncUserInfo, action, hybridMiniapp;
            var _a, _b;
            return __generator(this, function (_c) {
                withUnionId = (_a = options.withUnionId, _a === void 0 ? false : _a), createUser = (_b = options.createUser, _b === void 0 ? true : _b);
                syncUserInfo = this._scope === 'snsapi_base' ? false : options.syncUserInfo || false;
                action = 'auth.signIn';
                hybridMiniapp = this._runtime === RUNTIME.WX_MP ? '1' : '0';
                return [2, this._request.send(action, {
                        appid: appid,
                        loginType: loginType,
                        hybridMiniapp: hybridMiniapp,
                        syncUserInfo: syncUserInfo,
                        loginCredential: code,
                        withUnionId: withUnionId,
                        createUser: createUser,
                    }).then(function (res) {
                        if (res.code) {
                            throw new Error("[".concat(getSdkName(), "][").concat(ERRORS.OPERATION_FAIL, "] failed login via wechat: ").concat(res.code));
                        }
                        if (res.refresh_token) {
                            return {
                                refreshToken: res.refresh_token,
                                accessToken: res.access_token,
                                accessTokenExpire: res.access_token_expire,
                            };
                        }
                        throw new Error("[".concat(getSdkName(), "][").concat(ERRORS.OPERATION_FAIL, "] action:getJwt not return refreshToken"));
                    })];
            });
        });
    };
    __decorate([
        catchErrorsDecorator({
            title: '跳转微信公众号授权失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().weixinAuthProvider().signInWithRedirect() 的语法或参数是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], WeixinAuthProvider.prototype, "signInWithRedirect", null);
    __decorate([
        catchErrorsDecorator({
            title: '微信公众号登录失败',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().weixinAuthProvider().getRedirectResult() 的语法或参数是否正确',
                '  2 - 当前环境是否开通了微信公众号登录授权',
                '  3 - 微信公众号的 AppId 与 AppSecret 配置是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], WeixinAuthProvider.prototype, "getRedirectResult", null);
    __decorate([
        catchErrorsDecorator({
            title: '获取微信重定向绑定结果',
            messages: [
                '请确认以下各项：',
                '  1 - 调用 auth().weixinAuthProvider().getLinkRedirectResult() 的语法或参数是否正确',
                '  2 - 当前环境是否开通了微信公众号登录授权',
                '  3 - 微信公众号的 AppId 与 AppSecret 配置是否正确',
                "\u5982\u679C\u95EE\u9898\u4F9D\u7136\u5B58\u5728\uFF0C\u5EFA\u8BAE\u5230\u5B98\u65B9\u95EE\u7B54\u793E\u533A\u63D0\u95EE\u6216\u5BFB\u627E\u5E2E\u52A9\uFF1A".concat(COMMUNITY_SITE_URL),
            ],
        }),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", Promise)
    ], WeixinAuthProvider.prototype, "getLinkRedirectResult", null);
    return WeixinAuthProvider;
}(base_1.AuthProvider));
function getWeixinCode() {
    return getQuery('code') || getHash('code');
}
